Fixed classpath of javac used in compiled runtime

lutovich opened the pull request:
Incorrect classpath prevented queries from working after `mvn exec:java` in server directory.
Javac, used for compilation of generated classes, knew only about bootclasspath and thus was not able to locate any neo4j classes. Bootclasspath contained only some maven jars.

lutovich committed:
Removed some printing to console in tests

lutovich commented:
@thobe please review this PR

thobe commented on community/codegen/src/main/java/org/neo4j/codegen/source/Configuration.java:
```diff
@@ -98,49 +108,18 @@ public boolean isSet( SourceCode flag )
-        }
-        if ( warningsHandlers.size() == 1 )
-        {
-            return warningsHandlers.get( 0 );
```
No longer using the only one if there is only one?

lutovich replied:
Single one should still be fine with `WarningsHandler.Multiplex` as it does foreach inside.
Made it like this to be able to use short `? :`.
Should I add it back?

thobe replied:
I like not having to create the additional object indirection if it isn't required.

thobe commented on community/codegen/src/main/java/org/neo4j/codegen/source/SourceCode.java:
```diff
@@ -61,6 +63,24 @@ public String toString()
+                    classpath.append( File.pathSeparator );
+                    classpath.append( classpathUrl.getPath() );
+                }
+            }
```
It seems strange to _both_ fetch the `java.class.path` property _and_ go through the class loader to add the paths.

It seems like either we should do one or the other, if we do both, we should probably convert it to an (ordered) set of paths before converting it to a string, so that we don't add things twice. Probably we should also canonicalise the paths.

thobe replied:
If we are extracting the classpath from the ClassLoader, perhaps we should extract it from the ClassLoader that the user passed in, and perhaps we should go through the entire ClassLoader hierarchy and not just the first ClassLoader.

lutovich replied:
Absence of `java.class.path` makes `CodeGenerationTest` fail while running from maven. This happens because `SourceCode.class.getClassLoader().getURLs()` includes only `surefirebooter.jar`. So `codegen/target` is missing and `java.class.path` brings it.

I agree that it is safer to canonicalise and de-duplicate paths.

thobe commented:
Mostly nice. I think we might have to improve the `fullClasspath()` method though to ensure that we get a de-duplicated fully canonicalised classpath.

lutovich committed:
Clean up of code gen configuration

lutovich committed:
Fixed classpath of javac used in compiled runtime

Incorrect classpath prevented queries from working after `mvn exec:java` in server directory.
Javac, used for compilation of generated classes, knew only about bootclasspath and thus was not able to locate any neo4j classes.
Bootclasspath contained only some maven jars.

lutovich commented:
@thobe classpath assembling improved

thobe commented on community/codegen/src/main/java/org/neo4j/codegen/source/SourceCode.java:
```diff
@@ -208,4 +214,64 @@ public void handle( List<Diagnostic<? extends JavaFileObject>> diagnostics )
+        }
+        catch ( IOException e )
+        {
+            throw new RuntimeException( "Filed to get canonical path for: '" + path + "'", e );
```
Spelling: "Failed"

lutovich replied:
Ah, this is embarrassing :)

thobe commented:
Looks good. The code for forming the classpath is pretty sizeable though. Perhaps we should break it out into a separate class.

lutovich commented:
@thobe yes, good point. fixed now

thobe commented on community/codegen/src/main/java/org/neo4j/codegen/source/ClasspathHelper.java:
```diff
@@ -0,0 +1,117 @@
+import java.util.LinkedHashSet;
+import java.util.Set;
+
+public final class ClasspathHelper
```
`public`? ok, sure, not super important, just feels like a strange thing to expose from this API.

thobe commented on community/codegen/src/main/java/org/neo4j/codegen/source/ClasspathHelper.java:
```diff
@@ -0,0 +1,117 @@
+    public static Set<String> javaClasspath()
+    {
+        String[] classpathElements = javaClasspathString().split( ":" );
+        Set<String> result = new LinkedHashSet<>( classpathElements.length, 1 );
```
Isn't `1` a very high load factor? The hash table of the set will suffer a high degree of collisions.

thobe commented:
Looks good.
